Realm/ObjectStore/impl/realm_coordinator.cpp

```diff
@@ -0,0 +1,427 @@
+////////////////////////////////////////////////////////////////////////////
+//
+// Copyright 2015 Realm Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+////////////////////////////////////////////////////////////////////////////
+
+#include "realm_coordinator.hpp"
+
+#include "async_query.hpp"
+#include "cached_realm.hpp"
+#include "external_commit_helper.hpp"
+#include "object_store.hpp"
+#include "transact_log_handler.hpp"
+
+#include <realm/commit_log.hpp>
+#include <realm/group_shared.hpp>
+#include <realm/lang_bind_helper.hpp>
+#include <realm/query.hpp>
+#include <realm/table_view.hpp>
+
+#include <cassert>
+
+using namespace realm;
+using namespace realm::_impl;
+
+static std::mutex s_coordinator_mutex;
+static std::map<std::string, std::weak_ptr<RealmCoordinator>> s_coordinators_per_path;
+
+std::shared_ptr<RealmCoordinator> RealmCoordinator::get_coordinator(StringData path)
+{
+    std::lock_guard<std::mutex> lock(s_coordinator_mutex);
+    std::shared_ptr<RealmCoordinator> coordinator;
+
+    auto it = s_coordinators_per_path.find(path);
+    if (it != s_coordinators_per_path.end()) {
+        coordinator = it->second.lock();
+    }
+
+    if (!coordinator) {
+        s_coordinators_per_path[path] = coordinator = std::make_shared<RealmCoordinator>();
+    }
+
+    return coordinator;
```

Comment:
This could be written more clearly as:

```c++
    auto& weak_coordinator = s_coordinators_per_path[path];
    if (auto coordinator = weak_coordinator.lock()) {
        return coordinator;
    }
    auto coordinator = std::make_shared<RealmCoordinator>();
    weak_coordinator = coordinator;
    return coordinator;
```